pkg/addonmanager/controllers/managementaddonconfig/controller.go

```diff
@@ -0,0 +1,314 @@
+	return func(obj interface{}) {
+		name, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("error to get accessor of object: %v", obj))
+			return
+		}
+
+		objs, err := c.clusterManagementAddonIndexer.ByIndex(byClusterManagementAddOnConfig, fmt.Sprintf("%s/%s/%s", gvr.Group, gvr.Resource, name))
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("error to get addons: %v", err))
+			return
+		}
+
+		for _, obj := range objs {
+			if obj == nil {
+				continue
+			}
+
+			cma := obj.(*addonapiv1alpha1.ClusterManagementAddOn)
+			key, _ := cache.MetaNamespaceKeyFunc(cma)
+			c.queue.Add(key)
+		}
+	}
+}
+
+func (c *clusterManagementAddonConfigController) indexByConfig(obj interface{}) ([]string, error) {
+	cma, ok := obj.(*addonapiv1alpha1.ClusterManagementAddOn)
+	if !ok {
+		return nil, fmt.Errorf("obj is supposed to be a ClusterManagementAddOn, but is %T", obj)
+	}
+
+	configNames := sets.New[string]()
+	for _, defaultConfigRef := range cma.Status.DefaultConfigReferences {
+		if defaultConfigRef.DesiredConfig == nil || defaultConfigRef.DesiredConfig.Name == "" {
+			// bad config reference, ignore
+			continue
+		}
+
+		configNames.Insert(getIndex(defaultConfigRef.ConfigGroupResource, *defaultConfigRef.DesiredConfig))
+	}
+
+	for _, installProgression := range cma.Status.InstallProgressions {
+		for _, configReference := range installProgression.ConfigReferences {
+			if configReference.DesiredConfig == nil || configReference.DesiredConfig.Name == "" {
+				// bad config reference, ignore
+				continue
+			}
+
+			configNames.Insert(getIndex(configReference.ConfigGroupResource, *configReference.DesiredConfig))
+		}
+	}
+
+	if configNames.Len() == 0 {
+		// no configs , ignore
+		return nil, nil
+	}
+
+	return configNames.UnsortedList(), nil
+}
+
+func (c *clusterManagementAddonConfigController) sync(ctx context.Context, syncCtx factory.SyncContext, key string) error {
+	_, addonName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		// ignore addon whose key is invalid
+		return nil
+	}
+
+	cma, err := c.clusterManagementAddonLister.Get(addonName)
+	if errors.IsNotFound(err) {
+		// addon cloud be deleted, ignore
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	cmaCopy := cma.DeepCopy()
+
+	if err := c.updateConfigSpecHash(cmaCopy); err != nil {
+		return err
+	}
+
+	return c.patchConfigReferences(ctx, cma, cmaCopy)
+}
+
+func (c *clusterManagementAddonConfigController) updateConfigSpecHash(cma *addonapiv1alpha1.ClusterManagementAddOn) error {
+
+	for i, defaultConfigReference := range cma.Status.DefaultConfigReferences {
+		if defaultConfigReference.DesiredConfig == nil || defaultConfigReference.DesiredConfig.Name == "" {
+			continue
+		}
+
+		var err error
+		cma.Status.DefaultConfigReferences[i].DesiredConfig.SpecHash, err = c.getConfigSpecHash(defaultConfigReference.Group,
```

Comment:
I think we should directly set specHash, but check whether there is err, Otherwise this specHash will be set an empty string. I know we return error after that, but we should avoid directly setting here.